### PR TITLE
Use generic `TreeNodeTransformer` for forest-based node ID transformers

### PR DESCRIPTION
**Pull request by grovduck**
This PR introduces a base class for transformers that capture tree node IDs during `transform`.  Currently, this is only `RFNodeTransformer`, but #96 proposes other transformers with nearly duplicate needed functionality (gradient boosting and histogram-based gradient boosting).  This represents an incremental step to introducing those transformers and their accompanying NN-based estimators by doing the following:

- Creating a new abstract `TreeNodeTransformer` class that implements `transform` and `fit_transform`, but requires subclasses to implement `fit`.  This is because each specialized transformer will have estimator-specific keyword arguments and base forest types (e.g. `RandomForestClassifier` and `RandomForestRegressor`).  However, this class implements the method `_fit` which derived classes call with their specialized parameters and which sets the base estimator on each `y` target.
- Pulling the logic from `RFNodeTransformer` into `TreeNodeTransformer` for determining which base estimator to use for each `y` target as implemented in #91.
- Creating generic names to accommodate new transformers (e.g. `rfs_` to `estimators_` and `rf_type_dict_` to `estimator_type_dict_`)
- Rewriting tests to make clear the responsibility of both `TreeNodeTransformer` and `RFNodeTransformer`.

### CODE REVIEW

**Comment by grovduck** [comment]
@aazuspan, this one looks bigger than it actually is - lots of cutting and pasting from `RFNodeTransformer` to `TreeNodeTransformer`.  Locally, I've got the `GBNodeTransformer` already built and tested, but wanted to get this one as a separate PR before bringing that in.  

Once this one looks OK, I'll probably do the same abstraction on the NN estimator (as you recommended in #96) and then we can either do the user-overrides or start fleshing out the gradient boosting transformers/estimators.

File: tests/test_transformers.py [hunk 1]
```diff
@@ -233,41 +280,17 @@ def test_rfnode_transformer_non_default_parameterization(
     ).fit(X, y)
 
     # Check that both regression and classification forests are present
-    assert len(list(v == "classification" for v in est.rf_type_dict_.values())) >= 1
-    assert len(list(v == "regression" for v in est.rf_type_dict_.values())) >= 1
+    assert (
+        len(list(v == "classification" for v in est.estimator_type_dict_.values())) >= 1
+    )
+    assert len(list(v == "regression" for v in est.estimator_type_dict_.values())) >= 1
```
[the remark this answers is not in the document]
Comment (grovduck) [reply tests/test_transformers.py]: D'oh!

File: src/sknnr/transformers/_tree_node_transformer.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    def transform(self, X):
+        check_is_fitted(self)
+        _validate_data(
+            self,
+            X=X,
+            reset=False,
+            ensure_min_features=1,
+            ensure_min_samples=1,
+        )
+        return np.hstack([est.apply(X) for est in self.estimators_]).astype("int64")
```
[the remark this answers is not in the document]
Comment (grovduck) [reply src/sknnr/transformers/_tree_node_transformer.py:138]: @aazuspan, my thinking is that we will be weighting the results of running Hamming distance for each target, so this would happen at the level of the NN estimator and not the transformer.  In the case of `RFNNRegressor`, we are currently passing `metric=hamming` in the initializer.  I think we should be able to replace this with a custom [DistanceMetric](https://scikit-learn.org/0.24/modules/generated/sklearn.neighbors.DistanceMetric.html) that takes the weights assigned to each tree and calculates a weighted distance.  It looks like we can pass custom keyword arguments to this metric via the `metric_params` keyword in `KNeighborsRegressor`.

I think the one possibly tricky thing is that we will need to fit the underlying estimators (as you've [shown](https://github.com/lemma-osu/sknnr/issues/96#issuecomment-2967847215)) in order to get the tree weights.  We can set this as an estimator attribute (e.g. `tree_weights_`) during `fit`, but that will be too late to pass that object to the `KNeighborsRegressor` initializer.  Possibly, we pass a function that gets called which can retrieve the estimator attribute assuming it's been fit before neighbor finding?  I may be overcomplicating things here!

Now I haven't yet _done_ any of this 😉 , but that was my fuzzy path forward.
